### DISCUSSION TITLE
2. TCP/IP 4계층

### DISCUSSION THREAD

**Discussion by 24tngus** [24tngus/CS_STUDY#30] [Network]
#### 1. TCP(Transmission Control Protocol) / IP (Internet Protocol)

- 1) TCP/IP 4계층 개념
  - 장치간 인터넷을 통해 데이터 송수신에 사용되는 프로토콜 
  - IP로 패킷 전송시 TCP를 활용해 패킷을 정상적으로 수신 받음
  - 계층별 처리 역할이 달라서 계층별 간섭⬇️, 유지보수⬆️
  - 다른 계층간 데이터 전달 과정을 알 필요가 없어 데이터 캡슐화 가능 
![](https://velog.velcdn.com/images/24tngus/post/31bbcf9b-a4f9-4788-8420-45a1d5f39775/image.png)

- 2) 캡슐화/비캡슐화
  - 캡슐화 : 송신자가 수신자에게 데이터 보낼 때 계층을 지나면서 헤더 추가되는 과정
  - 🗒️ 예시 : 전송계층 ➡️ TCP 헤더 추가, 네트워크 계층 ➡️ IP 주소 헤더 추가 
  - 비캡슐화 : 수신자측에서 캡슐화된 데이터를 역순으로 제거하는 과정
  ![](https://velog.velcdn.com/images/24tngus/post/5f570d90-46b3-4339-a7ce-f79422e65977/image.png)
 
> 📖 참고 📖  PDU (Protocol Data Unit)
- 애플리케이션 계층 : 메시지
- 전송 계층 : 세그먼트(TCP), 데이터그램(UDP)
  - 세그먼트 : 적절한 크기로 쪼갠 조각 (세그먼트 = 데이터그램)
- 인터넷 계층 : 패킷
  - 패킷 : 세그먼트 SP와 DP가 포함된 IP 헤더가 붙은 형태의 조각
- 링크 계층 : 프레임(데이터 링크 계층), 비트(물리 계층)
  - 프레임 : MAC 주소 헤더와 CRC/체크섬 트레일러가 붙은 조각

> 📖 참고 📖 CRC/체크섬 트레일러
- 데이터 오류 감지를 위한 함수가 적용된 값
- 링크의 오류(과도한 트래픽)로 인해 데이터 손상을 감지하는 역할

#### 2. MTU, MSS, PMTUD

- MTU (Maximum Transmission Unit)
  - 네트워크 통신이 가능한 가장 큰 PDU의 크기
  - IP 헤더와 TCP 헤더 크기 포함 
  - MTU 확인 : `ping www.google.com -f -l 1500`
  ![](https://velog.velcdn.com/images/24tngus/post/6741a617-3fca-4428-877c-df0571dae3f5/image.png)
    - 1500바이트를 보냄
    - DF, 패킷분할이 안되서 손실률 100%
    - ping의 경우 IP 헤더(20바이트) + ICMP 헤더(8바이트)로 요청을 보내는 것 (MSS = 1472(1500 - 28)까지 가능)
  - 내 장치 MTU 확인 : (mac) `ifconfig` en0 (windows)`netsh interface ipv4 show interfaces` 
![](https://velog.velcdn.com/images/24tngus/post/ca973ed0-7e70-4d4c-8bdf-ae90c09a0319/image.png)
  
- MSS (Maximum Segment Size)
  - TCP에서 사용할 수 있는 데이터의 크기 
  - IP 헤더와 TCP 헤더 크기 포함X
  - 🗒️ 예시 : MTU 1500바이트여도 데이터는 1460바이트 이하 크기 보내야 전달 가능
  ![](https://velog.velcdn.com/images/24tngus/post/1f7df45a-5e83-4875-907c-e3cfca77728a/image.png)
  - 이더넷 프레임 : 이더넷 헤더 14바이트 + FCS 4바이트가 포함한 것
  - FCS (Frame Check Sequence) : 데이터 에러검출을 위해 삽입되는 필드 (CRC에 의해 생성된 것 포함)
  - 총 이더넷 프레임 크기 = 1518 바이트
  
- PMTUD (Path MTU Discovery)
  - 수신자와 송신자의 경로 상에서 장치가 패킷을 누락한 경우 패킷의 크기를 낮추면서 MTU에 맞게끔 반복해서 보내는 과정

- 패킷이 분할되지 않는 경우
  - 패킷을 분할할 수 없어 네트워크 경로 상에 있는 어떤 라우터나 장치의 MTU 초과하면, 아예 전달하지 않음
  - IPv6 : 분할X
  - IPv4 : flags 필드에서 bit=1이면, "Don't Fragment(플래그 활성화함)" 분할 불가능
  ![](https://velog.velcdn.com/images/24tngus/post/60e4b719-6bbb-4ffd-8b85-bdbc9ae0c4a3/image.png)

#### 3. 애플리케이션 계층

- 1) application layer 개념
  - 웹 서비스, 이메일 등 서비스를 사용자에게 실질적으로 제공
  - 🗒️ 예시 : HTTP, SMTP, SSH, FTP

- ** 2) HTTP (HyperText Transfer Protocol)**
  - 서버-브라우저간 (서버-서버간) 데이터를 주고 받기 위한 프로토콜
  - 헤더를 통한 확장 쉬움 (헤더값에 특정값 넣어 HTTP 요청시 쉽게 다른 값 추가 가능)
  - stateless (연속적인 상태값 없음)
  
- **3) SSH (Secure SHell)**
  - 보안되지 않는 네트워크간 서비스를 안전하게 운영하기 위한 암호화 네트워크 프로토콜
  - ssh <pem\> <user\>@<serverIP\>
  - scp 이용해서 SSH로 파일 전송 가능
  - scp <source\> <destinatino\>
  
- **4) FTP (File Transfer Protocol)**
  - 노드간 파일 전송시 사용되는 프로토콜
  - 현재는 파일 암호화 전송을 위해 **FTPS / SFTP** 사용
  - FTP 소프트웨어 : FileZilla
  
- **5) SMTP (Simple Mail Transfer Protocol)**
  - 인터넷을 통한 메일 전송 프로토콜
  - 서비스 운영시 node.js를 통해 메일 보낼 경우 사용
  - 🗒️ 예시 : 자바스크립트 `Nodemailer` 라이브러리 이용하여 JS 기반으로 SMTP 통해 메일 
```js
// create reusable transporter object using the default SMTP
transport
let transporter = nodemailer.createTransport({
host: "smtp.ethereal.email",
port: 587,
secure: false, // true for 465, false for other ports auth: {
user: testAccount.user, // generated ethereal user
pass: testAccount.pass, // generated ethereal password },
});
```
   
#### 4. 전송 계층

- 1) transport layer 개념
  - 메시지를 세그먼트(TCP)/데이터그램(UDP)으로 나눠 데이터를 오류 없이 순서대로 전달
  - 🗒️ 예시 :TCP, UDP
![](https://velog.velcdn.com/images/24tngus/post/ca9dd535-5a72-4ee8-902a-1d4f53007281/image.png)

- **2) TCP (Transmission Control Protocol)**
  - 가상회선 패킷 교환 방식 (순서 보장)
  - 전이중, 점대점 방식 (멀티캐스팅X, 브로드캐스팅X)
  - 오류 검사 : 재전송, 체크섬
  - 헤더 : 20 ~ 60바이트로 가변적
  ![](https://velog.velcdn.com/images/24tngus/post/a789d3d2-0752-4eb6-a436-a05ba8046204/image.png)

- **3) UDP (User Datagram Protocol)**
  - 데이터그램 패킷 교환 방식
  - 흐름제어X, 오류제어X, 재전송X
  - 오류 검사 : 체크섬만 지원
  - 헤더 : 8바이트 고정 길이
  - 🗒️ 예시 : DNS
  ![](https://velog.velcdn.com/images/24tngus/post/70bc98bd-4d20-40fe-a3cf-674f5fa73fe7/image.png)
  
> 📖 참고 📖 오류 검사 메커니즘
- ✏️ 재전송
  - 시간 초과 기간 지나면 서버는 전달되지 않은 데이터에 대해 재전송 시도
- ✏️ 체크섬
  - 송신된 데이터 체크섬과 수신된 체크섬 값을 비교해서 동일한지 확인 (무결성 평가)
  
#### 5. 인터넷 계층

- 1) Internet layer 개념
  - 세그먼트/데이터그램을 패킷화하여 한 노드에서 다른 노드(목적지)로 전송
  - 🗒️ 예시 : IP, ICMP, ARP, RARP
  
- **2) ICMP (Internet Control Message Protocol)**
  - 노드간 통신 확인 프로토콜
  - 데이터 교환하는데 사용X (테스팅에 사용)
  - 독립적인 비연결 프로토콜 
  ![](https://velog.velcdn.com/images/24tngus/post/8cfbc365-49cf-4eec-920b-471ce88d670e/image.png)
  
- **3) ARP** 
  - IP 주소 ➡️ MAC 주소 변환
  - 해당 주소에 맞는 MAC 주소 찾기 위해 Brodcasting 통해 연결된 네트워크 장치에 데이터 전송
  - 맞는 장치 있으면 해당 장치가 Unicast를 통해 데이터 전달해서 주소 찾음
  
- **4) RARP**
  - MAC 주소 ➡️ IP 주소 변환
  ![](https://velog.velcdn.com/images/24tngus/post/8626ea04-069a-48b6-b43e-42b495ec9e3a/image.png)

#### 6. 링크 계층 (네트워크 접속 계층)

- Link layer 개념
  - 데이터가 네트워크를 통해 물리적으로 전송
  - 🗒️ 예시 : 전선, 광섬유, 무선
  
#### 7. 3-way handshake (연결 성립)

- (1) SYN 단계 🤵‍♀️ ➡️ 🖥️
  - 클라이언트🤵‍♀️  : 서버에 클라이언트의 `ISN`을 담아 `SYN` 패킷 보냄
  - SYN (synchronization) : 연결 요청 플래그
- (2) SYN + ACK 단계 🖥️ ➡️ 🤵‍♀️ 
  - 서버🖥️ : 클라이언트의 SYN 수신하고, 서버의 `ISN`을 보내며 승인번호로 `ISN + 1` 보냄
- (3) ACK 단계 🤵‍♀️ ➡️ 🖥️
  - 클라이언트🤵‍♀️  : 서버의 `ISN + 1`한 값인 승인번호를 담아 ACK를 서버에 보냄
  - ACK (acknowledgement) : 응답 플래그
![](https://velog.velcdn.com/images/24tngus/post/a929601b-6daf-4037-a62b-570e246faa93/image.png)

- 클라이언트와 서버의 상태
  - 클라이언트🤵‍♀️ : closed, syn_sent, establised
  - 서버🖥️ : closed, listen, syn_received, established
  ![](https://velog.velcdn.com/images/24tngus/post/d1d681b4-5b41-4b6f-a359-86100bb34fa4/image.png)

> 📖 참고 📖 listen
- 서버가 클라이언트 연락 대기 상태 
- 해당 상태기반으로 서버 메서드 이름 결정
```java
const express = require('express') const app = express()
const port = 3000
app.get('/', (req, res) => {
  res.send('Hello World!')
})
app.listen(port, () => {
  console.log(`Example app listening on port ${port}`)
})
```

> 📖 참고 📖 ISN 
- TCP 기반 데이터 통신에서 각 새 연결에 할당된 고유한 32비트 시퀀스 번호
- TCP 연결을 통해 전송되는 다른 데이터와 충돌되지 않도록 시퀀스 번호 할당
- wireshark(패킷 분석 tool)로 TCP 연결 확인

#### 8. 4-way handshake (연걸 해제)

- (1) FIN 🤵‍♀️ ➡️ 🖥️
  - 클라이언트🤵‍♀️ : 서버에 FIN 설정 세그먼트 전송 후 FIN_WAIT_1 상태
  - 서버 응답 대기
- (2) ACK 🖥️ ➡️ 🤵‍♀️ 
  - 서버🖥️ : ACK라는 승인 세그먼트 전송 후 CLOSE_WAIT 상태
  - 클라이언트는 세그먼트 받으면 FIN_WAIT_2 상태
- (3) FIN 🖥️ ➡️ 🤵‍♀️ 
  - 서버🖥️ : LAST_ACK 상태가 되고, 일정 시간 이후 클라이언트에 FIN 세그먼트 전송 
- (4) ACK 🤵‍♀️ ➡️ 🖥️
  - 클라이언트🤵‍♀️ : TIME_WAIT 상태가 되고, 서버로 ACK 전송
  - 서버는 ACK 받고 CLOSED 상태
  - 클라이언트는 일정 시간(TIME_WAIT 설정 시간) 대기 후 연결 닫힘
  ![](https://velog.velcdn.com/images/24tngus/post/e4653ff8-0c0f-4dbe-b345-50bc0d40373c/image.png)

> 📖 참고 📖 TIME_WAIT
- 지연 패킷 등 발생시 데이터 무결성을 해결하기 위한 패킷 대기 시간
  - (FIN 플래그가 ACK 보다 먼저 도착한 경우)아직 서버로부터 받지 못한 데이터가 있을 것을 대비해 패킷 기다리는 과정
  - 2 * `MSL` 동안 대기
  - `MSL (Maximum Segment Lifetime)` : 최대 패킷 수명 (보통 2분)
- 연결이 올바르게 닫힌 상태로 만들기 위해 존재
  - CLOSED가 안되어 있을 경우 오류 발생하여, 일정 시간만 기다리도록 설정
  - CentOS/Ubuntu : 60초, Window :4분
  - Ubuntu 설정 파일
```c
  include/net/tcp.h
#define TCP_TIMEWAIT_LEN (60*HZ) //TIME-WAIT
```
![](https://velog.velcdn.com/images/24tngus/post/105637d9-8a2a-4ca3-895f-3c4569665e3e/image.png)

📖 참고 📖 [TCP / IP 과정](https://wooono.tistory.com/507)
- 1) 웹 브라우저에 `www.google.com` 입력
  - 구글 웹 서버에 80번 포트로 **HTTP Request** 보내는 것과 동일
- 2) **HTTP Request** 보내기 위해 각 계층에 필요한 정보를 패킷에 담아 전송
- 3) Application 계층 : **HTTP Request**  헤더 추가
![](https://velog.velcdn.com/images/24tngus/post/d4912c32-45d6-4b63-95a5-8bd87cecced6/image.png)
- 4) TCP 계층 : TCP 헤더
  - SP (출발지 포트 번호) : 내 컴퓨터에서 만든 소켓의 포트 번호
  - DP (목적지 포트 번호) : 80포트로 요청
![](https://velog.velcdn.com/images/24tngus/post/5afa8861-989e-4682-a13d-4acd48dfc575/image.png)
- 5) Internet 계층 : IP 헤더
  - SA (출발지 IP 주소) : 내 IP 주소
  - DA (목적지 IP 주소) : 도메인 정보로 IP주소 알기 위해 도메인 서버에 DNS 쿼리 보내고, IP 주소 응답 받음
![](https://velog.velcdn.com/images/24tngus/post/326a8ade-6c0f-4e58-ba75-0553918a3882/image.png)
- 6) Network Access 계층 : Ethernet 헤더
  - SA (출발지 MAC 주소) : 내 MAC 주소
  - DA (목적지 MAC 주소) : 구글 MAC 주소가 아닌, 물리적으로 연결된 패킷 전달된 라우터/게이트웨이의 MAC 주소 (ARP 프로토콜 사용)
![](https://velog.velcdn.com/images/24tngus/post/8441a409-a405-46d2-b520-2ad1df13f834/image.png)
- 7) 패킷 전송 전, 3-way handshaking으로 연결 설정
  - 컨트롤 비트(ACK, SYN 플래그) 사용
![](https://velog.velcdn.com/images/24tngus/post/b38f471d-434a-4417-9ba9-7d91cbf0ad85/image.png)![](https://velog.velcdn.com/images/24tngus/post/e847fef8-04f5-4817-9add-e0a31e7054df/image.png)
- 8) 라우팅 통해 완성된 패킷을 목적지 서버에 전송
  - `www.google.com`에 패킷 전송 후 HTML 응답 받음
  - 클라이언트는 응답 받은 HTML을 브라우저에 띄움
  ![](https://velog.velcdn.com/images/24tngus/post/26fb17cb-92fc-4f97-bfb2-b8c180882659/image.png)
- 9) HTTP 응답 끝나면, 4-way handshaking으로 연결 종료
  - 컨트롤 비트 (ACK, FIN 플래그) 사용
![](https://velog.velcdn.com/images/24tngus/post/cbec1def-5430-46c6-ba07-770943c866b5/image.png)![](https://velog.velcdn.com/images/24tngus/post/cf3c3dd7-f262-4e04-af7c-c2aaf85ff78e/image.png)

**Comment by mingkyeongg** [comment]
강의에서는 윈도우 기준으로 실습을 진행하여 제 컴퓨터와 호환이 되지 않는 경우가 많았는데 맥 기준으로도 다시 정리해주셔서 좋았습니다!!

**Comment by sujin-park0607** [comment]
확인했습니다! 면접 단골질문인 tcp 4계층 이쁘게 정리해주셔 자주 찾아볼것같아요! 감사합니다

**Comment by jindream6128** [comment]
확인했습니다 !!
마지막 TCP IP과정 찰떡이네요....

**Comment by berryberries** [comment]
확인했습니다!
